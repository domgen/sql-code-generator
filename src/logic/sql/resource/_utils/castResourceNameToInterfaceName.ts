import { pascalCase } from 'pascal-case';

import { ResourceType } from '../../../../model';

export const castResourceNameToInterfaceName = ({
  name,
  resourceType,
}: {
  name: string;
  resourceType: ResourceType;
}) => {
  const resourceTypeInTitleCase = resourceType[0].toUpperCase() + resourceType.substr(1).toLowerCase();
  const resourceNameInPascalCase = pascalCase(name);
  return `Sql${resourceTypeInTitleCase}${resourceNameInPascalCase}`;
};
