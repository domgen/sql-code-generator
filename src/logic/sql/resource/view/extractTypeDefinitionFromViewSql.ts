import { extractTypeDefinitionFromQuerySql } from '../../query/extractTypeDefinitionFromQuerySql';
import { TypeDefinitionOfResourceView } from '../../../../model/valueObjects/TypeDefinitionOfResourceView';
/*
  note: a view is effectively an alias for a query
    - with the condition that the query can not have input variables

  therefore, the types are a subset of the query types
*/
export const extractTypeDefinitionFromViewSql = ({ name, sql }: { name: string; sql: string }) => {
  // 0. grab the query definition from the view; e.g.: `CREATE VIEW __NAME__ AS SELECT ....` => `SELECT ...` => typedef
  const querySql = sql
    .split(/(SELECT|select)/)
    .slice(1) // slice(1) since part[0] = CREATE VIEW ... AS, part[1] = SELECT, part[2] = __SELECT_BODY__
    .join(''); // merge to have SELECT __SELECT_BODY__
  const queryDef = extractTypeDefinitionFromQuerySql({ sql: querySql });

  // 1. check that query def does not have any inputs, as views with inputs in query are invalid
  if (false) throw new Error(`query def for view '${name}' can not have inputs`); // TODO: when inputs are defined on query

  // 2. grab the selectExpressions and tableReferences off of the query
  return new TypeDefinitionOfResourceView({
    name,
    selectExpressions: queryDef.selectExpressions,
    tableReferences: queryDef.tableReferences,
  });
};
