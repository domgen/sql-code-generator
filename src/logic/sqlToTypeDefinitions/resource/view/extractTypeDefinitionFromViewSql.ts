import { TypeDefinitionOfResourceView } from '../../../../domain/objects/TypeDefinitionOfResourceView';
import { extractTypeDefinitionFromQuerySql } from '../../query/extractTypeDefinitionFromQuerySql';

/*
  note: a view is effectively a named alias for a query
    - with the condition that the query can not have input variables

  therefore, the types are a subset of the query types
*/
export const extractTypeDefinitionFromViewSql = ({
  name,
  sql,
}: {
  name: string;
  sql: string;
}) => {
  // 0. grab the query definition from the view; e.g.: `CREATE VIEW __NAME__ AS SELECT ....` => `SELECT ...` => typedef
  const querySql = sql
    .split(/(SELECT|select)/)
    .slice(1) // slice(1) since part[0] = CREATE VIEW ... AS, part[1] = SELECT, part[2] = __SELECT_BODY__
    .join(''); // merge to have SELECT __SELECT_BODY__
  const queryDef = extractTypeDefinitionFromQuerySql({
    name,
    path: 'none', // we dont currently track paths for view - and it wont be a problem downstream to fake it, so put a placeholder value
    sql: querySql,
  });

  // 1. check that query def does not have any inputs, as views with inputs in query are invalid
  if (queryDef.inputVariables.length)
    throw new Error(`query def for view '${name}' can not have inputs`);

  // 2. grab the selectExpressions and tableReferences off of the query
  return new TypeDefinitionOfResourceView({
    name,
    selectExpressions: queryDef.selectExpressions,
    tableReferences: queryDef.tableReferences,
  });
};
