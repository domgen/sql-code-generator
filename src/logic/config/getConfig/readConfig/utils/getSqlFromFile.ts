import { readFileAsync } from '../../../_utils/readFileAsync';

export const getSqlFromFile = async ({ filePath }: { filePath: string }): Promise<string> => {
  const extension = filePath.split('.').slice(-1)[0];
  if (extension === 'sql') return readFileAsync({ filePath });
  if (extension === 'ts') {
    const exports = await import(filePath);
    if (!exports.sql) throw new Error('.ts file did not have an export with name "sql"');
    return exports.sql;
  }
  throw new Error('path must lead to file with extension of .ts or .sql');
};
