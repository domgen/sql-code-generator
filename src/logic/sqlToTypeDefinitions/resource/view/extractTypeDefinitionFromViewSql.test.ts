import { extractTypeDefinitionFromViewSql } from './extractTypeDefinitionFromViewSql';
import { getSqlFromFile } from '../../../config/_utils/getSqlFromFile';

describe('extractTypeDefinitionFromViewSql', () => {
  it('should be able to extract types in this example', async () => {
    const exampleSql = await getSqlFromFile({
      filePath: `${__dirname}/../../../__test_assets__/views/view_suggestion_current.sql`,
    });
    const typeDef = extractTypeDefinitionFromViewSql({ name: 'view_suggestion_current', sql: exampleSql });
    expect(typeDef).toMatchSnapshot();
  });
  it.todo('should throw an error if query has input variables');
});
