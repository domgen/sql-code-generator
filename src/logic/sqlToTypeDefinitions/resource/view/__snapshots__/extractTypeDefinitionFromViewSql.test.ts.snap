// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`extractTypeDefinitionFromViewSql should be able to extract types in this example 1`] = `
TypeDefinitionOfResourceView {
  "name": "view_suggestion_current",
  "selectExpressions": Array [
    TypeDefinitionOfQuerySelectExpression {
      "alias": "id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "uuid",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.uuid",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "suggestion_source",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.suggestion_source",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "external_id",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.external_id",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "status",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.status",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "result",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.result",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "created_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "s.created_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "effective_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.effective_at",
      },
    },
    TypeDefinitionOfQuerySelectExpression {
      "alias": "updated_at",
      "typeReference": TypeDefinitionReference {
        "functionReferencePath": null,
        "tableReferencePath": "v.created_at",
      },
    },
  ],
  "tableReferences": Array [
    TypeDefinitionOfQueryTableReference {
      "alias": "s",
      "tableName": "suggestion",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "c",
      "tableName": "suggestion_cvp",
    },
    TypeDefinitionOfQueryTableReference {
      "alias": "v",
      "tableName": "suggestion_version",
    },
  ],
}
`;
